src/clj/rems/db/user_settings.clj

```diff
@@ -25,42 +26,42 @@
       validate-user-settings
       json/generate-string))
 
-(s/defschema PartialUserSettings
-  (map-keys s/optional-key UserSettings))
-
-(def ^:private coerce-partial-user-settings
-  (coerce/coercer! PartialUserSettings json/coercion-matcher))
+(def ^:private coerce-user-settings
+  (coerce/coercer! UserSettings json/coercion-matcher))
 
 (defn- json->settings [json]
-  ;; Allows missing keys, so we don't need to write migrations
-  ;; if we add new keys to user settings. Migrations are needed if
-  ;; we remove keys or make their validation stricter.
   (when json
     (-> json
         json/parse-string
-        coerce-partial-user-settings)))
+        coerce-user-settings)))
 
 (defn get-user-settings [user]
   (merge (default-settings)
          (when user
            (json->settings (:settings (db/get-user-settings {:user user}))))))
 
-(defn validate-new-settings [{:keys [language notification-email] :as settings}]
-  (into {} [(when (contains? (set (:languages env)) language)
-              {:language language})
-            (when (and notification-email (re-matches +email-regex+ notification-email))
-              {:notification-email notification-email})
-            (when (and (contains? settings :notification-email)
-                       (str/blank? notification-email))
-              ;; clear notification email to use identity provider's email instead
-              {:notification-email nil})]))
+(defn validate-new-settings [{:keys [language notification-email] :as new-settings}]
+  (let [valid-new-settings (merge {} ; valid returns at least a map
+                                  (when (contains? (set (:languages env)) language)
+                                    {:language language})
+                                  (when (and notification-email (re-matches +email-regex+ notification-email))
+                                    {:notification-email notification-email})
+                                  (when (and (contains? new-settings :notification-email)
+                                             (str/blank? notification-email))
+                                    ;; clear notification email to use identity provider's email instead
+                                    {:notification-email nil})
+                                  (when-let [ega (:ega new-settings)]
+                                    {:ega (select-keys ega [:api-key-expiration-date])}))]
+    (if (= (set (keys valid-new-settings))
```

Comment:
[golf] consider `when`

Reply:
If you agree that it's explicit enough then I'll do it.

Reply:
I kind of don't like the need to have an explicit `{}` in the beginning.

Reply:
I think we use the when pattern for other validation-like things a lot, so it's fine. I don't mind the `if` if you want to emphasize the `nil`.